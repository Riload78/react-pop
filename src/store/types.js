export const AUTH_LOGIN_PENDING = 'auth/login/pending'
export const AUTH_LOGIN_FULFILLED = 'auth/login/fulfilled'
export const AUTH_LOGIN_REJECTED = 'auth/login/rejected'
export const AUTH_LOGOUT = 'auth/logout'

export const SESSION_SAVE = 'session/save'

export const NOTIFICATION_CLOSE = 'notification/close'


export const ADVERTS_GET = 'adverts/get'
export const ADVERTS_GET_PENDING = 'adverts/get/pending'
export const ADVERTS_GET_FULFILLED = 'adverts/get/fulfilled'
export const ADVERTS_GET_REJECTED = 'adverts/get/rejected'
export const ADVERTS_POST = 'adverts/post'
export const ADVERTS_DELETE = 'adverts/delete'
export const ADVERTS_GET_TAGS = 'adverts/getTags'
