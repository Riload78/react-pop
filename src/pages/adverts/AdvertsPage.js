import getAdverts from './service.js'
import formatPrice from '../../helper/formatPrice.js'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Card from 'react-bootstrap/Card'
import Badge from 'react-bootstrap/Badge'
import Stack from 'react-bootstrap/Stack'
import Spinner from 'react-bootstrap/Spinner'
import { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'

const AdvertsPage = () => {
  const [adverts, setAdverts] = useState([])
  const [isLoading, setIsLoading] = useState(false)

  useEffect(() => {
    const ads = async () => {
      try {
        setIsLoading(true)
        const adverts = await getAdverts()
        setAdverts(adverts)
        setIsLoading(false)
      } catch (error) {
        setIsLoading(false)
        throw new Error(error.message)
      }
    }

    ads()
  }, [])

  return (
    <Container>
      {isLoading ? (
        <Row className=' min-vh-100 d-flex justify-content-center align-items-center'>
          <Spinner animation='border' variant='primary' size='3rem' />
        </Row>
      ) : (
        <Row xs={1} sm={2} md={3} lg={3} className='g-4'>
          {adverts.map(advert => (
            <Col key={advert.id}>
              <Card>
                <Link to={`${advert.id}`}>
                  <Card.Img variant='top' src={advert.photo} />
                </Link>
                <Card.Body>
                  <Link to={`${advert.id}`}>
                    <Card.Title>{advert.name}</Card.Title>
                  </Link>
                  <Card.Text className='d-flex justify-content-between align-items-center'>
                    {formatPrice(advert.price)}{' '}
                    <span>{advert.sale ? 'Venta' : 'Compra'}</span>
                  </Card.Text>
                </Card.Body>
                <Card.Footer>
                  <Stack direction='horizontal' gap={2}>
                    {advert.tags.map(tag => (
                      <Badge key={advert.id} pill bg='primary'>
                        {tag}
                      </Badge>
                    ))}
                  </Stack>
                </Card.Footer>
              </Card>
            </Col>
          ))}
        </Row>
      )}
    </Container>
  )
}

export default AdvertsPage
