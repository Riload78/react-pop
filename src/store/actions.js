import auth from '../pages/login/service.js'
import dataAdvert from '../pages/adverts/service.js'
import {
  isAdvertsLoaded,
  getAdvert,
  isAdvertDetailLoaded,
} from './selectors.js'
import {
  AUTH_LOGOUT,
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_FULFILLED,
  AUTH_LOGIN_REJECTED,
  SESSION_SAVE,
  ADVERTS_GET_PENDING,
  ADVERTS_GET_FULFILLED,
  ADVERTS_GET_REJECTED,
  ADVERTS_POST_PENDING,
  ADVERTS_POST_FULFILLED,
  ADVERTS_POST_REJECTED,
  ADVERTS_DELETE,
  ADVERTS_GET_TAGS,
  NOTIFICATION_CLOSE,
  ADVERTS_DETAIL_PENDING,
  ADVERTS_DETAIL_FULFILLED,
  ADVERTS_DETAIL_REJECTED,
} from './types'

export const authLogin = (credentials, isSessionSaved) => {
  return async function (dispatch) {
    try {
      dispatch(authLoginPending())
      await auth.login(credentials, isSessionSaved)
      dispatch(
        authLoginFulfilled({ type: 'success', message: 'LOGIN SUCCESSFUL' })
      )
    } catch (error) {
      dispatch(authLoginRejected({ type: 'error', message: error.message }))
    }
  }
}

export const authLoginPending = () => ({ type: AUTH_LOGIN_PENDING })

export const authLoginFulfilled = success => ({
  type: AUTH_LOGIN_FULFILLED,
  payload: success,
  notification: {
    type: success.type,
    message: success.message,
  },
})

export const authLoginRejected = error => ({
  type: AUTH_LOGIN_REJECTED,
  payload: error,
  notification: {
    type: error.type,
    message: error.message,
  },
})

export const authLogout = () => ({ type: AUTH_LOGOUT })

export const sessionSave = isSaved => ({
  type: SESSION_SAVE,
  payload: isSaved,
})

/* export const advertsLoad = adverts => ({
  type: ADVERTS_GET,
  payload: adverts,
}) */

// ADVERTS
export const advertsPending = () => ({
  type: ADVERTS_GET_PENDING,
})

export const advertsFulfilled = adverts => ({
  type: ADVERTS_GET_FULFILLED,
  payload: adverts,
})

export const advertsRejected = error => ({
  type: ADVERTS_GET_REJECTED,
  payload: error,
  notification: {
    type: error.type,
    message: error.message,
  },
})



// ADVERT POST
export const advertPostPending = () => ({
  type: ADVERTS_POST_PENDING,
})

export const advertPostFulfilled = advert => ({
  type: ADVERTS_POST_FULFILLED,
  payload: advert,
})

export const advertPostRejected = error => ({
  type: ADVERTS_POST_REJECTED,
  payload: error,
  notification: {
    type: error.type,
    message: error.message,
  },
})



export const createAdvert = advert => {
  return async function (dispatch) {
    try {
      dispatch(advertPostPending())
      const {id}= await dataAdvert.postAdvert(advert)
      const newAdvert = await dataAdvert.getAdvert(id)
      console.log('newAdvert:', newAdvert)
      dispatch(advertPostFulfilled(newAdvert))
      return newAdvert
    } catch (error) {
      dispatch(advertPostRejected({ type: 'error', message: error.message }))
    }
  }
}
export const advertDeleteFulfilled = id => ({
  type: ADVERTS_DELETE,
  payload: id,
})

export const advertDeleteRejected = error => ({
  type: ADVERTS_POST_REJECTED,
  payload: error,
  notification: {
    type: error.type,
    message: error.message,
  },
})

export const advertDelete = id => {
  return async function (dispatch) {
    try {
      const advert = await dataAdvert.getAdvert(id)
      const response = await dataAdvert.deleteAdvert(id)
      if(response.status === 204) {
        dispatch(
          advertDeleteRejected({ type: 'error', message: 'Advert not found' })
        )
      }
      dispatch(advertDeleteFulfilled(advert))
    } catch (error) {
      dispatch(advertDeleteRejected({ type: 'error', message: error.message }))
    }
  }
}


export const getTags = tags => ({
  type: ADVERTS_GET_TAGS,
  payload: tags,
})

export const advertsLoad = () => {
  return async function (dispatch, getState) {
    const state = getState()
    console.log('state:', state)
    console.log(isAdvertsLoaded(getState()))
    if (isAdvertsLoaded(state)) {
      console.log('adverts already loaded')
      return
    }
    try {
      dispatch(advertsPending())
      const adverts = await dataAdvert.getAdverts()
      dispatch(advertsFulfilled(adverts))
    } catch (error) {
      dispatch(advertsRejected({ type: 'error', message: error.message }))
    }
  }
}

// ADVERT
export const advertDetailPending = () => ({
  type: ADVERTS_DETAIL_PENDING,
})

export const advertDetailFulfilled = advert => ({
  type: ADVERTS_DETAIL_FULFILLED,
  payload: advert,
})

export const advertDetailRejected = error => ({
  type: ADVERTS_DETAIL_REJECTED,
  payload: error,
  notification: {
    type: error.type,
    message: error.message,
  },
})

export const advertLoad = advertId => {
  return async function (dispatch, getState) {
    const state = getState()
    if (isAdvertDetailLoaded(state)) {
      return
    }
    try {
      dispatch(advertDetailPending())
      const advertDetail = await dataAdvert.getAdvert(advertId)
      dispatch(advertDetailFulfilled(advertDetail))
    } catch (error) {
      dispatch(advertDetailRejected({ type: 'error', message: error.message }))
    }
  }
}

export const hideNotification = () => ({
  type: NOTIFICATION_CLOSE,
  payload: null,
  notification: {
    type: null,
    message: null,
  },
})
