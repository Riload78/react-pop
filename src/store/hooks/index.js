import UseAdvert from "./UseAdvert";
import UseAdverts from "./UseAdverts";
import UseAdvertDelete from "./UseAdvertDelete";
import UseCreateAdvert from "./UseCreateAdvert";
import UseTags from "./UseTags";  

const hooks = {
    UseAdvert,
    UseAdverts,
    UseAdvertDelete,
    UseCreateAdvert,
    UseTags
}

export default hooks