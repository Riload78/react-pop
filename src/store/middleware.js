export const logger = store => next => action => {
  console.group(action.type)
  console.info('dispatching', action, store.getState())
  const result = next(action)
  console.log('final state', store.getState())
  console.groupEnd()
  return result
}
