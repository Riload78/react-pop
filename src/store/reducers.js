import { combineReducers } from 'redux'
import {
  AUTH_LOGOUT,
  AUTH_LOGIN,
  SESSION_SAVE,
  ADVERTS_GET,
  ADVERTS_POST,
  ADVERTS_DELETE,
  ADVERTS_GET_TAGS,
} from './types'

export const defaultState = {
  auth: false,
  session: true,
  adverts: [],
}

export const auth = (state = defaultState.auth, action) => {
  switch (action.type) {
    case AUTH_LOGIN:
      return true
    case AUTH_LOGOUT:
      return false
    default:
      return state
  }
}

export const session = (state = {}, action) => {
  switch (action.type) {
    case SESSION_SAVE:
      return true
    default:
      return state
  }
}

export const adverts = (state = defaultState.adverts, action) => {
  switch (action.type) {
    case ADVERTS_GET:
      return action.payload
    case ADVERTS_POST:
      return [...state.adverts, action.payload]
    case ADVERTS_DELETE:
      return [
        ...state.adverts.filter(advert => advert.id !== action.payload.id),
      ]
    default:
      return state
  }
}

const reducer = combineReducers({
  auth,
  session,
  adverts,
})

export default reducer
