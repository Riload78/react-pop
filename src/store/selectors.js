export const getIsLogin = state => state.auth
export const getIsLogout = state => !state.auth

export const getIsLoading = state => state.ui.loading

export const getIsSaved = state => state.session


export const getAdverts = state => state.adverts
export const getAdvert = advertId => state =>
  getAdverts(state).find(advert => advert.id === advertId)

export const getTags = state => state.adverts.tags
