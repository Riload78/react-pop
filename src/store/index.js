import { combineReducers, createStore, applyMiddleware } from 'redux'
import { composeWithDevTools } from '@redux-devtools/extension'
import * as reducers from './reducers'
import * as actionCreators from './actions'
import { thunk } from 'redux-thunk'

const reducer = combineReducers(reducers)

const composeEnhancers = composeWithDevTools({ actionCreators })

export default function configureStore(preloadedState) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(thunk))
  )

  return store
}
