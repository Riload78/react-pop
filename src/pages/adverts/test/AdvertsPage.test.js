import { render, screen } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import AdvertsPage from '../AdvertsPage'
import { Provider } from 'react-redux'
import { advertsLoad } from '../../../store/actions'

jest.mock('../../../store/actions')

describe('AdvertsPage', () => {
  const state = {
    adverts: {
      isloadeded: true,
      data: [],
      maxPrice: 0,
    },
    ui: {
      loading: false,
      notification: {
        type: null,
        message: null,
      },
    },
  }

  const store = {
    dispatch: () => {},
    getState: () => state,
    subscribe: () => {},
    replaceReducer: () => {},
  }

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <Router>
          <AdvertsPage />
        </Router>
      </Provider>
    )

  it('should render without crashing', () => {
    const { container } = renderComponent()
    expect(advertsLoad).toHaveBeenCalled()
    expect(container).toMatchSnapshot()
  })

  it('should render Not Result if Advert is empty', () => {
    state.adverts.data = []
    renderComponent()
    expect(
      screen.getByText('There are no publications yet')
    ).toBeInTheDocument()
  })

  it('should return a list of Adverts', () => {
    state.adverts.data = [
      {
        createdAt: '2024-06-21T12:37:04.000Z',
        id: '76e68c8c-1185-47b0-b55c-759d40cbd587',
        name: 'Editado',
        photo: null,
        price: 123,
        sale: false,
        tags: ['mobile'],
      },
    ]
    renderComponent()
    const advertsList = screen.getAllByRole('list')
    expect(screen.queryByText('There are no publications yet')).not.toBeInTheDocument()
				expect(advertsList).toHaveLength(1)
  })
})
