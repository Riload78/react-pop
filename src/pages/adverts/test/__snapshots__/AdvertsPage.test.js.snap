// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvertsPage should render without crashing 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="login-wrapper d-flex justify-content-center align-items-center w-100 container"
    >
      <div
        class=" d-flex justify-content-center align-self-center g-4 text-center row"
      >
        <h4>
          There are no publications yet
        </h4>
        <h3>
          Do you want to create a new post?
        </h3>
        <button
          class="w-auto btn btn-primary btn-xl"
          type="button"
        >
          Create
        </button>
      </div>
    </div>
  </div>
</div>
`;
